### PR TITLE
Fix piping more than one CommandInfo to Get-Command

### LINKED ISSUE #10851: Get-Command doesn't recognize System.Management.Automation.CommandInfo (its own output type) as pipeline input
Feeding `Get-Command` output to itself via the pipeline - useful if you want to force loading of all modules in the context of reflection - doesn't work as expected.

It looks like a new parameter set with a `System.Management.Automation.CommandInfo`-typed parameter  would have to be introduced to support that.

# Steps to reproduce

```powershell
# OK
('Add-Content', 'Get-Content' | Get-Command).Name | Should -Be 'Add-Content', 'Get-Content'

# Fails
(Get-Command 'Add-Content', 'Get-Content' | Get-Command).Name | Should -Be 'Add-Content', 'Get-Content'
```

# Expected behavior

Both tests should succeed.

# Actual behavior

The 2nd test fails, because the only  `CommandInfo` object output is the one for `Add-Content`

```none
Expected @('Add-Content', 'Get-Content'), but got 'Add-Content'
```

# Environment data

```none
PowerShell Core 7.0.0-preview.4
```

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

Piping multiple CommandInfo to `Get-Command` only returns the first result.  This is because during discovery of the command it internally builds a cache of commands it has already seen.  On the second (or more) times `ProcessRecord()` is called, that cache is already filled with some results.  If the requested CommandInfo (which binds to `-Verb` and `-Noun`) happens to be in that cache, then the cmdlet considers that a duplicate and doesn't return that result.  The fix in this case is to clear the cache at the start of `ProcessRecord()` to treat each search as independent.

## PR Context

Fix https://github.com/PowerShell/PowerShell/issues/10851

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/GetCommandCommand.cs [hunk 1]
```diff
@@ -390,6 +390,8 @@ protected override void BeginProcessing()
         /// </summary>
```
Comment (PoshChan) [inline src/System.Management.Automation/engine/GetCommandCommand.cs]: @SteveL-MSFT, your last commit had 4 failures in `PowerShell-CI-linux`
Breakpoint SDK Unit Tests.Managing breakpoints in a remote runspace via the SDK.Can enable breakpoints
```powershell
Exception calling "EnableBreakpoint" with "1" argument(s): "The background process reported an error with the following message: Remote transport error: ."
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1: line 106
106:                 $bp = $jobRunspace.Debugger.EnableBreakpoint($bp)                

```
Breakpoint SDK Unit Tests.Managing breakpoints in a remote runspace via the SDK.Can remove breakpoints
```powershell
Exception calling "GetBreakpoints" with "0" argument(s): "The runspace state is not valid for this operation."
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1: line 120
120:             foreach ($bp in $jobRunspace.Debugger.GetBreakpoints()) {

```
Breakpoint SDK Unit Tests.Managing breakpoints in a remote runspace via the SDK.Returns an empty collection when there are no breakpoints
```powershell
Exception calling "GetBreakpoints" with "0" argument(s): "The runspace state is not valid for this operation."
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1: line 126
126:             $jobRunspace.Debugger.GetBreakpoints() | Should -HaveCount 0

```
Breakpoint SDK Unit Tests.Handling errors while managing breakpoints in a remote runspace via the SDK.Error occurred in Context block
```powershell
Exception calling "SetCommandBreakpoint" with "1" argument(s): "The runspace state is not valid for this operation."
At /home/vsts/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1:153 char:13
```

**Comment by SteveL-MSFT** [comment]
@PoshChan please retry linux

**Comment by PoshChan** [comment]
@SteveL-MSFT, successfully started retry of `PowerShell-CI-Linux`

File: test/powershell/Modules/Microsoft.PowerShell.Core/Get-Command.Tests.ps1 [hunk 2]
```diff
@@ -263,4 +263,10 @@ Describe "Get-Command Tests" -Tags "CI" {
         VerifyDynamicParametersExist -cmdlet $results[0] -parameterNames $paramName
         VerifyParameterType -cmdlet $results[0] -parameterName $paramName -ParameterType System.Text.Encoding
     }
+
+    It "Piping more than one CommandInfo works" {
+        $result = Get-Command Add-Content, Get-Content | Get-Command
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Core/Get-Command.Tests.ps1]:
```suggestion
        $result = Get-Command -Name Add-Content, Get-Content | Get-Command
```

**Review by iSazonov — APPROVED** [review]

**Review by rjmholt — APPROVED** [review]

**Review by adityapatwardhan — APPROVED** [review]

**Comment by mklement0** [comment]
Unfortunately, this PR didn't fix all problems, because the following tests fail:

```
function foo {}; set-alias foo foo

(gcm -all foo | gcm).Count | Should -Be 2

(gcm -all foo | gcm).ComandType | Should -Be Alias, Function
```

**Comment by SteveL-MSFT** [comment]
@mklement0 those are different issues not related to piping to Get-Command, please open a new issue for them

**Comment by mklement0** [comment]
@SteveL-MSFT: These tests _do_ pipe to `Get-Command` (whose built-in alias is `gcm`). In what way is this a different issue?

**Comment by mklement0** [comment]
If it helps with getting this resolved, please see #11017

**Comment by SteveL-MSFT** [comment]
@mklement0 my mistake, I only saw the left side of the pipe

**Pull request merged**